### PR TITLE
[Feature] binding method allow support can execute

### PR DESCRIPTION
**Pull request by workgroupengineering**
## What does the pull request do?

To simplify the use of the Commands in the MVVM methodology in AvaloniaUI without defining a property with type ICommand when it is necessary that the command execution depends on the state of the ViewModel.

## What is the current behavior?

When Binding Command of Button(MenuItem,..) to method, the button it is always enabled

## What is the updated/expected behavior with this PR?

The Enable of Button(MenuItem,...) depends of  state of ViewModel

## How was the solution implemented (if it's not obvious)?

The example of how it works can be found in [unit test](https://github.com/workgroupengineering/Avalonia/blob/cf60b5484e915732b6e763f75ca26745e3c8384a/tests/Avalonia.Markup.Xaml.UnitTests/Data/BindingTests_Method.cs#L118 ) .
The engine finds in the ViewModel method which has the same name as the association of the method with the prefix Can.
Example:
if you associate the method with the name Do engine it finds the function with the name CanDo with an input argument of type object and return a Boolean value. If the engine does not find the function, execution is always allowed.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
none

## Fixed issues
none

### CODE REVIEW

**Comment by grokys** [comment]
I've thought about implementing this in the past and it feels like a good idea to me! Hopefully it won't take anyone by surprise?

It will definitely need some documentation over at https://github.com/AvaloniaUI/Avaloniaui.net when it's ready though.

**Review by workgroupengineering — COMMENTED** [review]
> 
> 
> This uses a lot of reflection and some runtime codegen. We'll need to think of ways to do this with compiled bindings as well for a zero-reflection path.

@jkoritzinsky  Can you give me how codegen works in AvaloniaUI? where can i find the documentation?

@jkoritzinsky Do you have an idea why when CanExecuteChanged is called CommandParameter it hasn't changed yet?

**Comment by workgroupengineering** [comment]
I have aligned the PR to the master branch, corrected the test ,updated the BindingDemo example, but I have a problem with ReactiveObject.
When the ViewModel is derived from AlwaysEnabledDelegateCommand's ReactiveObject [invalidateCanExecuteHandler](https://github.com/AvaloniaUI/Avalonia/pull/4095/files#diff-a8024c79d992346ebc447193cfa2a282R57-R65) it is called only the first time.

**Comment by workgroupengineering** [comment]
> 
> 
> I have aligned the PR to the master branch, corrected the test ,updated the BindingDemo example, but I have a problem with ReactiveObject.
> When the ViewModel is derived from AlwaysEnabledDelegateCommand's ReactiveObject [invalidateCanExecuteHandler](https://github.com/AvaloniaUI/Avalonia/pull/4095/files#diff-a8024c79d992346ebc447193cfa2a282R57-R65) it is called only the first time.

I solved it, it was enough to update to the latest version of ReactiveUI

File: samples/BindingDemo/BindingDemo.csproj [hunk 1]
```diff
@@ -15,4 +15,7 @@
   <Import Condition="'$(TargetFramework)'=='net461'" Project="..\..\build\NetFX.props" />
   <Import Project="..\..\build\ReferenceCoreLibraries.props" />
   <Import Project="..\..\build\BuildTargets.targets" />
+  <ItemGroup>
+    <PackageReference Update="ReactiveUI" Version="11.4.17" />
```
Comment (rstm-sf) [inline samples/BindingDemo/BindingDemo.csproj:19]: Hello!

Shouldn't we also update the following?

https://github.com/AvaloniaUI/Avalonia/blob/70a0f00422b6638492e1bbf3c7de847bc8ac01c9/build/ReactiveUI.props#L3

File: src/Avalonia.Base/Data/Converters/AlwaysEnabledDelegateCommand.cs [hunk 2]
```diff
@@ -8,34 +10,174 @@ namespace Avalonia.Data.Converters
 {
     class AlwaysEnabledDelegateCommand : ICommand
```
[the remark this answers is not in the document]
Comment (workgroupengineering) [reply src/Avalonia.Base/Data/Converters/AlwaysEnabledDelegateCommand.cs]: I renamed AlwaysEnabledDelegateCommand to CommandToMethodsConverter

File: src/Avalonia.Base/Data/Converters/AlwaysEnabledDelegateCommand.cs [hunk 3]
```diff
@@ -8,34 +10,174 @@ namespace Avalonia.Data.Converters
...
+                {
+                    System.ComponentModel.PropertyChangedEventHandler invalidateCanExecuteHandler = (s, e) =>
+                    {
+                        if (string.IsNullOrWhiteSpace(e.PropertyName)
+                            || dependencyProperties.Contains(e.PropertyName))
+                        {
+                            CanExecuteChanged?.Invoke(this, EventArgs.Empty);
+                        }
+                    };
+                    inpc.PropertyChanged += invalidateCanExecuteHandler;
```
[the remark this answers is not in the document]
Comment (workgroupengineering) [reply src/Avalonia.Base/Data/Converters/AlwaysEnabledDelegateCommand.cs]: Implemented WeakPropertyChangedProxy to avoid memory leak

**Comment by workgroupengineering** [comment]
@jkoritzinsky  Can you review?

**Comment by jkoritzinsky** [comment]
I'll take a look later this week or this weekend.

**Review by jkoritzinsky — APPROVED** [review]
LGTM! We'll want to add support for this to compiled bindings when we add support for the method->command binding.

**Pull request merged**